dask_ml/feature_extraction/text.py

```diff
@@ -108,3 +110,114 @@ class FeatureHasher(_BaseHasher, sklearn.feature_extraction.text.FeatureHasher):
     @property
     def _hasher(self):
         return sklearn.feature_extraction.text.FeatureHasher
+
+
+class Vocabulary:
+    vocabulary = {}
+    fixed_vocabulary = None
+
+    def __init__(self, vocabulary=None):
+        if vocabulary is None:
+            self.vocabulary = set()
+            self.fixed_vocabulary = False
+        else:
+            self.fixed_vocabulary = True
+            self.vocabulary = vocabulary
+
+    def update(self, obj):
+        assert not self.fixed_vocabulary
+        self.vocabulary |= obj
+
+    def finalize(self):
+        if self.fixed_vocabulary:
+            return self.vocabulary
+        else:
+            return {key: i for i, key in enumerate(sorted(self.vocabulary))}
+
+    @property
+    def n_features(self):
+        return len(self.vocabulary)
+
+
+class CountVectorizer(sklearn.feature_extraction.text.CountVectorizer):
+    # Potential issues:
+    # 1. The vocabularies might be somewhat large.
+    #    Consider storing this on the cluster, rather than shipping between
+    #    the client?
+    def fit_transform(self, raw_documents, y=None):
+        # Just bag for now.
+        # Two cases:
+        # 1. vocabulary provided, easy.
+        # 2. vocabulary learned, harder.
+        params = self.get_params()
+        vocabulary = params.pop("vocabulary")
+
+        client = get_client()
+        vocabulary_ = client.submit(Vocabulary, vocabulary=vocabulary, actor=True)
+        vocabulary_actor = vocabulary_.result()
+
+        if self.vocabulary is not None:
+            # Case 1: Just map transform.
+            fixed_vocabulary = True
+            result = raw_documents.map_partitions(
+                _count_vectorizer_transform, vocabulary_actor, params
+            )
+        else:
+            fixed_vocabulary = False
+            # Case 2: learn vocabulary from the data.
+            vocabularies = raw_documents.map_partitions(
+                _build_vocabulary, vocabulary_actor, params
+            )
+            # compute and merge.
+            dask.compute(vocabularies)  # List[Set[str]]
+            # same as before
+            result = raw_documents.map_partitions(
+                _count_vectorizer_transform, vocabulary_actor, params
+            )
```

Comment:
So if we're doing two passes anyway, why use actors here?  Why not do something like a tree reduction?  

I see in your original post there is some performance difference.  Can I ask what approach you tried previously?

Reply:
I pushed a commit making actors a toggle. https://github.com/dask/dask-ml/pull/705/files#diff-67ef0189a95e65d0765c18596857a748R269-R281 has the non-actor version. Roughly:

1. Build a vocabulary for each partition.
2. Collect all the vocabularies on the client.
3. Merge the vocabularies.

I don't immediately see where a tree reduction is helpful (but I don't think I've ever implemented one before). I think that our "tree" structure is very flat here: A `List[Set[str]]` that we need to merge into a single `Set[str]` with a set union.

I'll poke around the dask repo. I know I've seen references to them there.